### ISSUE TITLE
Infinite loop caused by _updateSubscriptions

### ISSUE THREAD

**Issue by Disperito** [solana-labs/solana#26198]
#### Problem
I'm using the version 1.44.1 of @solana/web3.js, to develop an application with high-frequency calling.
<img width="1196" alt="image" src="https://user-images.githubusercontent.com/50449082/175457437-5bcd9adf-8217-4bb3-a199-030fa268bb18.png">

**Comment by steveluscher** [comment]
Thanks for this report!

Other than spamming the error console, does this interfere with the operation of your app?

**Comment by Disperito** [comment]
> Thanks for this report!
> 
> Other than spamming the error console, does this interfere with the operation of your app?

Of course. It caused ***infinite recursion*** and made CPU usage soar to 100%, so that my app cannot run as normal.

**Comment by steveluscher** [comment]
Totally, I get it, @Disperito. Unbounded recursion is super bad news, no matter which way you slice it. What I was trying to get at is that there are two kinds of infinite recursion.

One leads to the JavaScript thread literally locking, and freezing the browser:

```
while (true) {
  // Browser's dead.
}
```

The other results in code running as often as it can, and high CPU usage, but doesn't lock the thread; other code can continue to run in the period between yields.

```
while (true) {
  // Between the time this yields and resumes, other work can get done on the JS thread.
  await someAsyncWork();
}
```

In any case, we need a general solution for this.

**Comment by steveluscher** [comment]
Can I ask you a question? Can you count the number of subscriptions that you're making to the RPC _in total_? I've seen situations where RPCs limit the number of open subscriptions you may hold. When you exceed a certain number, they murder your connection abruptly. See https://github.com/solana-labs/solana/issues/25322#issuecomment-1138168590.

**Comment by Disperito** [comment]
> @Disperito. Unbounded recursion is super bad news, no matter which way you slice it. What I was trying to get at is that there are two kinds of infinite recursion.
> 
> One leads to the JavaScript thread literally locking, and freezing the browser:

Sorry I have no definite answer. But I have reviewed the logs near when this error occurs, I think there were only less than 10 subscriptions at the same time, even less.

**Comment by Jac0xb** [comment]
Still getting this error with @solana/web3.js@1.56.2 @steveluscher.
<img width="321" alt="image" src="https://user-images.githubusercontent.com/5273873/189410313-30a3cd46-832f-4b28-9843-98694a3bec76.png">

```
signatureSubscribe error for argument [
'{SOME_SIGNATURE}',
{ commitment: 'confirmed' }
] socket not ready
```

**Comment by steveluscher** [comment]
@Jac0xb If you have a reliable repro of this, can you debug it a bit to see if this is the root cause? https://github.com/solana-labs/solana/issues/25578

**Comment by Jac0xb** [comment]
Not sure the repro steps but happens often with sendConfirmation and possibly log lived connection objects. Haven't seen it before.

**Comment by steveluscher** [comment]
To anyone who's still experiencing this problem, can you try to upgrade to an experimental release, by running `yarn add @solana/web3.js@pr-29195`? That's a package built from PR #29195.

Let me know if that fixes anything.

**Comment by steveluscher** [comment]
@Jac0xb, @Disperito, @gallynaut do you have time to give this another try, if you still have a repro? See the message above.
